### ISSUE TITLE
KeyUp / KeyDown KeyEventHandler Queued

### ISSUE THREAD

**Issue by Tanjitsu** [crosire/scripthookvdotnet#646]
```
public KeyCapture()
{
      Tick += new EventHandler(OnTick);

      KeyDown += OnKeyDown;
}

OnKeyDown { Camera(); }

OnTick {  if (IsControlJustReleased) { Camera(); }   }

Camera
{
     Screen.Message("Start"); KeyCapture.Wait( 5000 ); Screen.Message("End");
}

```

The above being a very basic example.. not exact code just the layout :)

Pressing the Control pad button repeatedly will Start a message only if a Press occurs after the Wait. 
Pressing the Keyboard key will stack up requests to the KeyEventHandler.
Once the Wait releases they then begin to execute the queued commands even through the KeyPress was several seconds ago.

This is not the behaviour id expect so.... is this a Bug or is there a work around i can use to dispose of the events when the script is not active ?

**Comment by RainerHilmer** [comment]
This is normal behavior. You need to implement a key event filter.  Take the general purpose [key event filter](https://bitbucket.org/CyronsTeam/vcommonfunctions/src/e586d90a0bce7681ee07dff08ea8ad6a7d2cde3c/For_shvdn2/VCommonFunctions/KeyEventFilter.cs?at=master&fileviewer=file-view-default)  of my VCommonFunctions library.

**Comment by Tanjitsu** [comment]
@RainerHilmer Thank you for the reply. Although i cant see anything in [pastebin/PK1eeUE3](https://pastebin.com/PK1eeUE3) relating to Event handling ? :)

**Comment by RainerHilmer** [comment]
This code is just the filter. You have to place it in-between like in this demo:
`using System;
using System.Collections.Generic;
using System.Windows.Forms;
using Cyron43.GtaV.Common;

namespace KeyEventFilter.Tests
{
   internal class KeyHandlingDemo : Control
   {
      private readonly KeyEventFilter _keyEventFilter = new KeyEventFilter();
      private readonly SomeScript _someScript = new SomeScript();

      internal KeyHandlingDemo()
      {
         var keysToInclude = new List<KeyContainer>
                               {
                                  new KeyContainer { Key = Keys.D },
                                  new KeyContainer { Key = Keys.T }
                               };
         _keyEventFilter.IncludedKeys = keysToInclude;
         KeyUp += _keyEventFilter.FilterKeyEvent;
         _keyEventFilter.FilteredKeyEvent += (sender, e) => HandleKey(e);
      }

      private void AtDoThat(KeyEventArgs e)
      {
         if(e.KeyCode != Keys.D)
            return;
         _someScript.DoThat();
      }

      private void AtTriggerThis(KeyEventArgs e)
      {
         if(e.KeyCode != Keys.T)
            return;
         _keyEventFilter.FilterKeyEvent(this, e);
         _someScript.TriggerThis();
      }

      private void HandleKey(KeyEventArgs e)
      {
         AtTriggerThis(e);
         AtDoThat(e);
         AtChangeSpeed(e);
      }

      private void AtChangeSpeed(KeyEventArgs e)
      {
         //...
      }
   }

   internal class SomeScript //: Script
   {
      public void DoThat()
      {
         throw new NotImplementedException();
      }

      public void TriggerThis()
      {
         throw new NotImplementedException();
      }
   }
}`
In a nutshell, you attach a special eventhandler from the Filter to the Tick event. The filter comes with it's own event which you attach your eventhandler to. So the line of communication is Tick event -> Filter handler and then Filter-Tick event -> your handler. Here is a link to the complete help file of my VCommonFunctions library. https://drive.google.com/open?id=0B8vO0ixlizXvZmhjR1lMR3U2OUk

**Comment by Tanjitsu** [comment]
Thank you again RainerHilmer, i have added a simple solution that works :)

```
public class KeyFilter : Script
{
        public KeyFilter()
        {
            KeyDown += OnKey;
        }

        private void OnKey(object sender, KeyEventArgs e)
        {
            KeyFilter.Yield();

            if (e != null && !e.Handled) e.Handled = true;
        }
}
```

I can then handle the keyEvents as normal in my KeyCapture script but just check if (!e.Handled).
While my main script KeyCapture is waiting KeyFilter will flag any keyEvents as handled.

**Comment by RainerHilmer** [comment]
Thank you for telling about the Handled property. I didn't even know about it. LOL

**Comment by Tanjitsu** [comment]
I added this to the Wiki for future reference.
